storage/memorybysubnet/peer_store.go

```diff
@@ -0,0 +1,595 @@
+	}
+	shard.swarms[ih].seeders[mask][pk] = time.Now().UnixNano()
+
+	shard.Unlock()
+	return nil
+}
+
+func (s *peerStore) DeleteSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-s.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	shard := s.shards[s.shardIndex(ih, p.IP.AddressFamily)]
+	shard.Lock()
+
+	if _, ok := shard.swarms[ih]; !ok {
+		shard.Unlock()
+		return storage.ErrResourceDoesNotExist
+	}
+
+	mask := s.mask(p)
+	if _, ok := shard.swarms[ih].seeders[mask][pk]; !ok {
+		shard.Unlock()
+		return storage.ErrResourceDoesNotExist
+	}
+
+	delete(shard.swarms[ih].seeders[mask], pk)
+
+	if shard.swarms[ih].lenSeeders()|shard.swarms[ih].lenLeechers() == 0 {
+		delete(shard.swarms, ih)
+		recordInfohashesDelta(-1)
+	}
+
+	shard.Unlock()
+	return nil
+}
+
+func (s *peerStore) PutLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-s.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	shard := s.shards[s.shardIndex(ih, p.IP.AddressFamily)]
+	shard.Lock()
+
+	if _, ok := shard.swarms[ih]; !ok {
+		shard.swarms[ih] = swarm{
+			seeders:  make(map[string]map[serializedPeer]int64),
+			leechers: make(map[string]map[serializedPeer]int64),
+		}
+		recordInfohashesDelta(1)
+	}
+
+	mask := s.mask(p)
+	if shard.swarms[ih].leechers[mask] == nil {
+		shard.swarms[ih].leechers[mask] = make(map[serializedPeer]int64)
+	}
+	shard.swarms[ih].leechers[mask][pk] = time.Now().UnixNano()
+
+	shard.Unlock()
+	return nil
+}
+
+func (s *peerStore) DeleteLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-s.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	shard := s.shards[s.shardIndex(ih, p.IP.AddressFamily)]
+	shard.Lock()
+
+	if _, ok := shard.swarms[ih]; !ok {
+		shard.Unlock()
+		return storage.ErrResourceDoesNotExist
+	}
+
+	mask := s.mask(p)
+	if _, ok := shard.swarms[ih].leechers[mask][pk]; !ok {
+		shard.Unlock()
+		return storage.ErrResourceDoesNotExist
+	}
+
+	delete(shard.swarms[ih].leechers[mask], pk)
+
+	if shard.swarms[ih].lenSeeders()|shard.swarms[ih].lenLeechers() == 0 {
+		delete(shard.swarms, ih)
+		recordInfohashesDelta(-1)
+	}
+
+	shard.Unlock()
+	return nil
+}
+
+func (s *peerStore) GraduateLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-s.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	shard := s.shards[s.shardIndex(ih, p.IP.AddressFamily)]
+	shard.Lock()
+
+	if _, ok := shard.swarms[ih]; !ok {
+		shard.swarms[ih] = swarm{
+			seeders:  make(map[string]map[serializedPeer]int64),
+			leechers: make(map[string]map[serializedPeer]int64),
+		}
+		recordInfohashesDelta(1)
+	}
+
+	mask := s.mask(p)
+	delete(shard.swarms[ih].leechers[mask], pk)
```

Comment:
`shard.swarms[ih].leechers[mask]` can be nil if we just created `shard.swarms[ih]` above. Does `delete` panic if called on a nil map?